Sources/ApolloTestSupport/MockGraphQLServer.swift

```diff
@@ -0,0 +1,73 @@
+@testable import Apollo
+import XCTest
+
+public class MockGraphQLServer {
+  enum ServerError: Error, CustomStringConvertible {
+    case unexpectedRequest(String)
+    
+    public var description: String {
+      switch self {
+      case .unexpectedRequest(let requestDescription):
+        return "Mock GraphQL server received an unexpected request: \(requestDescription)"
+      }
+    }
+  }
+  
+  public typealias RequestHandler<Operation: GraphQLOperation> = (HTTPRequest<Operation>) -> JSONObject
+  
+  private class RequestExpectation<Operation: GraphQLOperation>: XCTestExpectation {
+    let file: StaticString
+    let line: UInt
+    let handler: RequestHandler<Operation>
+    
+    init(description: String, file: StaticString = #filePath, line: UInt = #line, handler: @escaping RequestHandler<Operation>) {
+      self.file = file
+      self.line = line
+      self.handler = handler
+
+      super.init(description: description)
+    }
+  }
+  
+  private let queue = DispatchQueue(label: "com.apollographql.MockGraphQLServer")
+    
+  public init() { }
+  
+  // Using existentials to avoid having to deal with type erasure.
+  private var requestExpectations: [AnyHashable: Any] = [:]
+  
+  private subscript<Operation: GraphQLOperation>(_ operationType: Operation.Type) -> RequestExpectation<Operation>? {
+    get {
+      requestExpectations[ObjectIdentifier(operationType)] as! RequestExpectation<Operation>?
```

Comment:
`ObjectIdentifier` is being used because the meta type (`.Type`) isn't hashable. From what I could find, that was a common workaround.

Reply:
You can only have one request expectation configured per operation type at a time, but you can configure it to be expected to invoked multiple times. And you can always override existing expectations for a later part of the test. Are you concerned you may want separate expectations for consecutive invocations, similar to `AsyncResultObserver`? I could add that, but do you have an example in mind?

Reply:
The thought was something like "call once, validate cache is populated, then call again to make sure result is updated in the cache" on the same operation, but with different results each time

Reply:
Ah yes, that's a pattern the current tests already use. Because these are separate steps, you can call `expect` again before the second step to update the expectation. That's what happens here for example: https://github.com/apollographql/apollo-ios/blob/7c1807d5a9402ed817b0dec1aec6865d04df2d4f/Tests/ApolloCacheDependentTests/WatchQueryTests.swift#L93-L153

Where it gets tricky is if control doesn't come back to the test before you want the expectation to change. I have no concrete scenario in mind, but that is definitely a limitation. Maybe it could be enough to pass in a count of the number of times the operation has been called, so you can use that return different results?

Reply:
🤔 yeah that could help - and then if on the 3rd call you're getting the 2nd result it makes it clearer why you're not getting what you expect

Reply:
> > Maybe it could be enough to pass in a count of the number of times the operation has been called, so you can use that return different results?

> 🤔 yeah that could help - and then if on the 3rd call you're getting the 2nd result it makes it clearer why you're not getting what you expect
> I think that could help here. 

Hmmm, I'm not actually sure how that would help with what you're describing. What are you suggesting here exactly? 

It should already be clear when the wrong expectation is being triggered, because you set them up in order, and they have different checks. For example, if you add a `fetchResultFromCacheExpectation` and a `fetchResultFromServerExpectation`, your assertion would fail when `fetchResultFromCacheExpectation` is triggered with a source of `server` first. (And if a 3rd calls comes in you would get an 'unexpected result' failure.)

> I'm also still wondering what `ObjectIdentifier` is 🙃

Did you miss https://github.com/apollographql/apollo-ios/pull/1482#discussion_r513828831 or did you have questions about it?

Reply:
Right I see that it's a workaround, I'm not sure what its origin is - is this something Foundation is doing or Swift or ??

I suppose you're right about the wrong result (theoretically) causing failure anyway.

Reply:
It's a current Swift language limitation. Metatypes can't conform to protocols like `Hashable`.

Reply:
Right that part I understand. I mean where does the `ObjectIdentifier(object)`...function? type?...come from?

Reply:
Oh, `ObjectIdentifier` is part of the Swift standard library: https://developer.apple.com/documentation/swift/objectidentifier

Reply:
Ahhhhhhh ok - for some reason when I was trying to click through to docs it wasn't going through to the stdlib. Thank you!